### PR TITLE
feat: mechanism to let Gardener configure OS settings

### LINKED ISSUE #310: Enable cgroups v2
**What would you like to be added**:

With https://github.com/gardenlinux/gardenlinux/pull/205 we have removed cgroup v2 support for compatibility reasons. This should be enabled again.

**Why is this needed**:

A bunch of interesting features depend on it such as Rootless Kubernetes and more precise memory accounting useful for resource reservations.

### PR DESCRIPTION
**Pull request by MrBatschner**
**How to categorize this PR?**

/kind enhancement
/area os
/os garden-linux

**What this PR does / why we need it**:
Introduces a set of scripts and systemd units that allow Gardener to declaratively change certain OS related settings. Currently these include changing the cgroup hierarchy, the default Linux Security Module (AppArmor or SELinux) and the netfilter frontend (iptables-legacy or iptables-nft).

**Which issue(s) this PR fixes**:
Fixes #310

**Release note**:

```feature user
feat: mechanism to let Gardener configure OS settings
feat: configuration and reboot mimic
feat: provide a script to reconfigure cgroup
feat: scripts to configure Linux Security Module
feat: configuration script for netfilter frontend
```

### CODE REVIEW

**Review by Vincinator — COMMENTED** [review]
Hi, some comments on the code below.

**General Question**
We have an asynchrony between writing the the desired configuration to the control files, and the system applying the configuration. The service `gardener-configure-settings.service` is a oneshot. 

In case of an error, should we write the status back somewhere? 
And maybe also mark the content in the control files as invalid/not-applied? 
Just in case some script or person might interpret the content as the current state.

Maybe this is to nitpicky though. 

Otherwise looks good to me.

File: features/gardener/file.include/var/lib/gardener-gardenlinux/restart_kernel.sh [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+if [ ! -e "$cmdline" ]; then
+    echo "kernel command line could not be found at $cmdline" >&2
+    exit 3
+fi
+
+# we seem to have everything so lets restart the kernel with kexec
+rm -f $RESTART_CONTROL_FILE
+
+# kexec would be nice because it is fast but at the moment, it seems to be failing on some
```
[the remark this answers is not in the document]
Comment (MrBatschner) [reply features/gardener/file.include/var/lib/gardener-gardenlinux/restart_kernel.sh:42]: Will do.

File: features/gardener/file.include/var/lib/gardener-gardenlinux/restart_kernel.sh [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    exit 2
+fi
+
+if [ ! -e "$cmdline" ]; then
+    echo "kernel command line could not be found at $cmdline" >&2
+    exit 3
+fi
+
+# we seem to have everything so lets restart the kernel with kexec
+rm -f $RESTART_CONTROL_FILE
```
[the remark this answers is not in the document]
Comment (MrBatschner) [reply features/gardener/file.include/var/lib/gardener-gardenlinux/restart_kernel.sh]: Done (along with other Shellcheck recommendations).

File: features/gardener/file.include/var/lib/gardener-gardenlinux/03_iptables_backend.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+if [ "$desired_nff" == "iptables" -o "$desired_nff" == "legacy" ]; then
+    nf_frontend="/usr/sbin/iptables-legacy"
+    nf6_frontend="/usr/sbin/ip6tables-legacy"
+elif [ "$desired_nff" == "nftables" -o "$desired_nff" == "nft" ]; then
+    nf_frontend="/usr/sbin/iptables-nft"
+    nf6_frontend="/usr/sbin/ip6tables-nft"
+fi
+
+for nff in $nf_frontend $nf_frontend; do
```
[the remark this answers is not in the document]
Comment (MrBatschner) [reply features/gardener/file.include/var/lib/gardener-gardenlinux/03_iptables_backend.sh]: Oh yes, thank you for catching that.

File: features/gardener/file.include/var/lib/gardener-gardenlinux/03_iptables_backend.sh [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        fi
+    fi
+    echo $mode
+}
+
+# no need to run if Gardener did not place a configuration for netfilter backend
+[ ! -f "$NETFILTER_GARDENER" ] && exit 0
+
+desired_nff=$(cat "$NETFILTER_GARDENER")
+current_nff=$(check_netfilter_frontend)
```
[the remark this answers is not in the document]
Comment (MrBatschner) [reply features/gardener/file.include/var/lib/gardener-gardenlinux/03_iptables_backend.sh]: Gone!

**Comment by MrBatschner** [comment]
> **General Question** We have an asynchrony between writing the the desired configuration to the control files, and the system applying the configuration. The service `gardener-configure-settings.service` is a oneshot.
> 
> In case of an error, should we write the status back somewhere? And maybe also mark the content in the control files as invalid/not-applied? Just in case some script or person might interpret the content as the current state.

The (ugly) truth is that at the moment, the Gardener bootstrapping process is a one-way route. It can run some scripts during bootstrapping (those which would populate the files in `/var/lib/gardener-gardenlinux/etc` but it cannot read back any status. Therefore, I did not pay any attention to it and will adapt the whole mimic here once that feature becomes available in Gardener.

**Review by Vincinator — APPROVED** [review]
/lgtm

**Comment by Vincinator** [comment]
> > **General Question** We have an asynchrony between writing the the desired configuration to the control files, and the system applying the configuration. The service `gardener-configure-settings.service` is a oneshot.
> > In case of an error, should we write the status back somewhere? And maybe also mark the content in the control files as invalid/not-applied? Just in case some script or person might interpret the content as the current state.
> 
> The (ugly) truth is that at the moment, the Gardener bootstrapping process is a one-way route. It can run some scripts during bootstrapping (those which would populate the files in `/var/lib/gardener-gardenlinux/etc` but it cannot read back any status. Therefore, I did not pay any attention to it and will adapt the whole mimic here once that feature becomes available in Gardener.

Ok, got it. I agree, to wait for the status/feedback/whatever feature to see how integration should work. Thanks for the input!

File: features/gardener/file.include/var/lib/gardener-gardenlinux/restart_kernel.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (MrBatschner) [reply features/gardener/file.include/var/lib/gardener-gardenlinux/restart_kernel.sh:42]: See issue #754.

**Pull request merged**